File: bindings/python/iota_sdk/types/unlock.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations
from dataclasses import dataclass, field
from enum import IntEnum
from typing import Dict, List, TypeAlias, Union, Any
from dataclasses_json import config
from iota_sdk.types.signature import Ed25519Signature
from iota_sdk.types.common import json


class UnlockType(IntEnum):
    """Unlock variants.

    Attributes:
        Signature (0): An unlock holding a signature unlocking one or more inputs.
        Reference (1): An unlock which must reference a previous unlock which unlocks also the input at the same index as this Reference Unlock.
        Account (2): An unlock which must reference a previous unlock which unlocks the account that the input is locked to.
        Anchor (3): An unlock which must reference a previous unlock which unlocks the anchor that the input is locked to.
        Nft (4): An unlock which must reference a previous unlock which unlocks the NFT that the input is locked to.
        Multi (5): Unlocks a MultiAddress with a list of other unlocks.
        Empty (6): Used to maintain correct index relationship between addresses and signatures when unlocking a MultiUnlock where not all addresses are unlocked.

    """
    Signature = 0
    Reference = 1
    Account = 2
    Anchor = 3
    Nft = 4
    Multi = 5
    Empty = 6


@json
@dataclass
class SignatureUnlock:
    """An unlock holding a signature unlocking one or more inputs.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Signature),
        init=False)
    signature: Ed25519Signature


@json
@dataclass
class ReferenceUnlock:
    """An unlock which must reference a previous unlock which unlocks also the input at the same index as this Reference Unlock.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Reference),
        init=False)
    reference: int


@json
@dataclass
class AccountUnlock:
    """An unlock which must reference a previous unlock which unlocks the account that the input is locked to.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Account),
        init=False)
    reference: int


@json
@dataclass
class AnchorUnlock:
    """An unlock which must reference a previous unlock which unlocks the anchor that the input is locked to.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Anchor),
        init=False)
    reference: int


@json
@dataclass
class NftUnlock:
    """An unlock which must reference a previous unlock which unlocks the NFT that the input is locked to.
    """
    type: int = field(default_factory=lambda: int(UnlockType.Nft), init=False)
    reference: int


# pylint: disable=missing-function-docstring,unused-argument
def deserialize_unlocks(dicts: List[Dict[str, Any]]) -> List[Unlock]:
    # Function gets overwritten further below, but needs to be defined here
    # already
    pass


@json
@dataclass
class MultiUnlock:
    """Unlocks a MultiAddress with a list of other unlocks.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Multi),
        init=False)
    unlocks: List[Unlock] = field(metadata=config(
        decoder=deserialize_unlocks
    ))


@json
@dataclass
class EmptyUnlock:
    """Used to maintain correct index relationship between addresses and signatures when unlocking a MultiUnlock where not all addresses are unlocked.
    """
    type: int = field(
        default_factory=lambda: int(
            UnlockType.Empty),
        init=False)


Unlock: TypeAlias = Union[SignatureUnlock,
                          ReferenceUnlock,
                          AccountUnlock,
                          AnchorUnlock,
                          NftUnlock,
                          MultiUnlock,
                          EmptyUnlock]

# pylint: disable=too-many-return-statements


def deserialize_unlock(d: Dict[str, Any]) -> Unlock:
    """
    Takes a dictionary as input and returns an instance of a specific class based on the value of the 'type' key in the dictionary.

    Arguments:
    * `d`: A dictionary that is expected to have a key called 'type' which specifies the type of the returned value.
    """
    unlock_type = d['type']
    if unlock_type == UnlockType.Signature:
        return SignatureUnlock.from_dict(d)
    if unlock_type == UnlockType.Reference:
        return ReferenceUnlock.from_dict(d)
    if unlock_type == UnlockType.Account:
        return AccountUnlock.from_dict(d)
    if unlock_type == UnlockType.Anchor:
        return AnchorUnlock.from_dict(d)
    if unlock_type == UnlockType.Nft:
        return NftUnlock.from_dict(d)
    if unlock_type == UnlockType.Multi:
        return MultiUnlock.from_dict(d)
    if unlock_type == UnlockType.Empty:
        return EmptyUnlock.from_dict(d)
    raise Exception(f'invalid unlock type: {unlock_type}')

# pylint: disable=function-redefined


def deserialize_unlocks(dicts: List[Dict[str, Any]]) -> List[Unlock]:
    """
    Takes a list of dictionaries as input and returns a list with specific instances of classes based on the value of the 'type' key in the dictionary.

    Arguments:
    * `dicts`: A list of dictionaries that are expected to have a key called 'type' which specifies the type of the returned value.
    """
    return list(map(deserialize_unlock, dicts))

File: bindings/nodejs/lib/types/block/payload/signed_transaction/unlock.ts
// Copyright 2023 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import { Type } from 'class-transformer';
import { Ed25519Signature } from '../../signature';

/**
 * All of the unlock types.
 */
enum UnlockType {
    /**
     * A signature unlock.
     */
    Signature = 0,
    /**
     * A reference unlock.
     */
    Reference = 1,
    /**
     *  An account unlock.
     */
    Account = 2,
    /**
     *  An Anchor unlock.
     */
    Anchor = 3,
    /**
     *  An NFT unlock.
     */
    Nft = 4,
    /**
     *  A multi unlock.
     */
    Multi = 5,
    /**
     *  An empty unlock.
     */
    Empty = 6,
}

/**
 * The base class for unlocks.
 */
abstract class Unlock {
    /**
     * The type of unlock.
     */
    readonly type: UnlockType;

    /**
     * @param type The type of unlock.
     */
    constructor(type: UnlockType) {
        this.type = type;
    }
}

/**
 * An unlock holding one or more signatures unlocking one or more inputs..
 */
class SignatureUnlock extends Unlock {
    /**
     * The signature.
     */
    @Type(() => Ed25519Signature)
    readonly signature: Ed25519Signature;

    /**
     * @param signature An Ed25519 signature.
     */
    constructor(signature: Ed25519Signature) {
        super(UnlockType.Signature);
        this.signature = signature;
    }
}

/**
 * An unlock which must reference a previous unlock which unlocks
 * also the input at the same index as this Reference Unlock.
 */
class ReferenceUnlock extends Unlock {
    /**
     * The reference.
     */
    readonly reference: number;

    /**
     * @param reference An index referencing a previous unlock.
     */
    constructor(reference: number) {
        super(UnlockType.Reference);
        this.reference = reference;
    }
}

/**
 * An unlock which must reference a previous unlock which unlocks the account that the input is locked to.
 */
class AccountUnlock extends Unlock {
    /**
     * The reference.
     */
    readonly reference: number;

    /**
     * @param reference An index referencing a previous unlock.
     */
    constructor(reference: number) {
        super(UnlockType.Account);
        this.reference = reference;
    }
}

/**
 * An unlock which must reference a previous unlock which unlocks the anchor that the input is locked to.
 */
class AnchorUnlock extends Unlock {
    /**
     * The reference.
     */
    readonly reference: number;

    /**
     * @param reference An index referencing a previous unlock.
     */
    constructor(reference: number) {
        super(UnlockType.Anchor);
        this.reference = reference;
    }
}

/**
 * An unlock which must reference a previous unlock which unlocks the NFT that the input is locked to.
 */
class NftUnlock extends Unlock {
    /**
     * The reference.
     */
    readonly reference: number;

    /**
     * @param reference An index referencing a previous unlock.
     */
    constructor(reference: number) {
        super(UnlockType.Nft);
        this.reference = reference;
    }
}

/**
 * Unlocks a MultiAddress with a list of other unlocks.
 */
class MultiUnlock extends Unlock {
    /**
     * The inner unlocks.
     */
    @Type(() => Unlock, {
        // @ts-ignore:next-line: no-use-before-declare
        discriminator: UnlockDiscriminator,
    })
    readonly unlocks: Unlock[];

    /**
     * @param unlocks The inner unlocks.
     */
    constructor(unlocks: Unlock[]) {
        super(UnlockType.Multi);
        this.unlocks = unlocks;
    }
}

/**
 * Used to maintain correct index relationship between addresses and signatures when unlocking a MultiUnlock where not all addresses are unlocked.
 */
class EmptyUnlock extends Unlock {
    constructor() {
        super(UnlockType.Empty);
    }
}

const UnlockDiscriminator = {
    property: 'type',
    subTypes: [
        {
            value: SignatureUnlock,
            name: UnlockType.Signature as any,
        },
        {
            value: ReferenceUnlock,
            name: UnlockType.Reference as any,
        },
        {
            value: AccountUnlock,
            name: UnlockType.Account as any,
        },
        {
            value: AnchorUnlock,
            name: UnlockType.Anchor as any,
        },
        {
            value: NftUnlock,
            name: UnlockType.Nft as any,
        },
        {
            value: MultiUnlock,
            name: UnlockType.Multi as any,
        },
        {
            value: EmptyUnlock,
            name: UnlockType.Empty as any,
        },
    ],
};

export {
    UnlockType,
    Unlock,
    SignatureUnlock,
    ReferenceUnlock,
    AccountUnlock,
    AnchorUnlock,
    NftUnlock,
    MultiUnlock,
    EmptyUnlock,
    UnlockDiscriminator,
};
